src/Raven.Client/Documents/Session/DocumentSession.Query.cs

```diff
@@ -24,7 +40,11 @@ public partial class DocumentSession
         /// <returns></returns>
         public IDocumentQuery<T> DocumentQuery<T, TIndexCreator>() where TIndexCreator : AbstractCommonApiForIndexes, new()
         {
-            var index = new TIndexCreator();
+            if (IndexMetadataCache.TryGetValue(typeof(TIndexCreator), out var index) == false)
```

Comment:
what about Async session? Maybe create internal static class called IndexMetadataCache with all required methods there?

Reply:
What do you mean? I've covered the async session, but the static class seems good.